kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/AccountManagerImpl.java

```diff
@@ -0,0 +1,167 @@
+	@Override
+	public boolean isAccountCreated() {
+		return local.getAccountState() == CREATION_COMPLETED;
+	}
+
+	@Override
+	public void addAccountStateObserver(@NonNull Observer<Integer> observer) {
+		accountState.addObserver(observer);
+		accountState.postValue(accountState.getValue());
+	}
+
+	@Override
+	public void removeAccountStateObserver(@NonNull Observer<Integer> observer) {
+		accountState.removeObserver(observer);
+	}
+
+	@Override
+	public void retry() {
+		if (kinAccount != null && accountState.getValue() == ERROR) {
+			setAccountState(local.getAccountState());
+		}
+	}
+
+	@Override
+	public void start(@NonNull final KinAccount kinAccount) {
+		this.kinAccount = kinAccount;
+		Logger.log(new Log().withTag(TAG).put("setAccountState", "start"));
+		if (getAccountState() != CREATION_COMPLETED) {
+			if (getAccountState() == ERROR) {
+				retry();
+			} else {
+				this.setAccountState(getAccountState());
+			}
+		}
+	}
+
+	private void setAccountState(@AccountState final int accountState) {
+		if (isValidState(this.accountState.getValue(), accountState)) {
+			if (accountState != ERROR) {
+				this.local.setAccountState(accountState);
+			}
+			this.accountState.postValue(accountState);
+			switch (accountState) {
+				case REQUIRE_CREATION:
+					eventLogger.send(StellarAccountCreationRequested.create());
+					Logger.log(new Log().withTag(TAG).put("setAccountState", "REQUIRE_CREATION"));
+					// Trigger account creation from server side.
+					authRepository.getAuthToken(new KinCallbackAdapter<AuthToken>() {
+						@Override
+						public void onResponse(AuthToken response) {
+							setAccountState(PENDING_CREATION);
+						}
+					});
+					break;
+				case PENDING_CREATION:
+					Logger.log(new Log().withTag(TAG).put("setAccountState", "PENDING_CREATION"));
+					// Start listen for account creation on the blockchain side.
+					accountCreationRegistration = kinAccount.blockchainEvents()
+						.addAccountCreationListener(new EventListener<Void>() {
+							@Override
+							public void onEvent(Void data) {
+								accountCreationRegistration.remove();
+								accountCreationRegistration = null;
+								setAccountState(REQUIRE_TRUSTLINE);
+							}
+						});
+					break;
+				case REQUIRE_TRUSTLINE:
+					Logger.log(new Log().withTag(TAG).put("setAccountState", "REQUIRE_TRUSTLINE"));
+					// Create trustline transaction with KIN
+					new CreateTrustLineCall(kinAccount, new TrustlineCallback() {
+						@Override
+						public void onSuccess() {
+							eventLogger.send(StellarKinTrustlineSetupSucceeded.create());
+							setAccountState(CREATION_COMPLETED);
+						}
+
+						@Override
+						public void onFailure(OperationFailedException e) {
+							eventLogger.send(StellarKinTrustlineSetupFailed.create(e.getMessage()));
+							setAccountState(ERROR);
+						}
+					}).start();
+					break;
+				case CREATION_COMPLETED:
+					// Mark account creation completed.
+					eventLogger.send(WalletCreationSucceeded.create());
```

Comment:
CREATION_COMPLETED shouldn't happen twice in any case. I think we need to protect it in isValidState()